### DISCUSSION TITLE
authExchange duplicating operations?

### DISCUSSION THREAD

**Discussion by jbalke** [urql-graphql/urql#1070] [Q&A] [answered]
I'm new to urql so not ruling out that I've made a mistake somewhere...

For background, I'm building a NextJS app with next-urql.

Initially, I implemented an authExchange (JWT) based on this gist: https://gist.github.com/kitten/6050e4f447cb29724546dd2e0e68b470. All worked fine. Some time after switching to the official exchange-auth package I noticed my backend receiving duplicate mutations (what tipped me off were the primary key violations on registering new users). Urql devtools has since confirmed that all of my queries/mutations are firing off in duplicate; one execution command resulting in two results within milliseconds of each other, all identical keys.

In a process of elimination, I discovered that the duplicate operations(?) only occur when using the new exchange-auth package. I'd post my authExchange config but it's practically identical to the documented example, other than minor implementation details such as how I get/set my access token, identify auth errors and log users out.

So I'm at a loss here. Is there anything I can do to debug further?

Btw, this is what my exchanges looks like: [devtoolsExchange, dedupExchange, cache, retryExchange(options), errorExchange, fetchOptions(ctx), authExchange({/* auth config */}), ssrExchange, fetchExchange].

**Comment by kitten** [comment]
That's pretty odd! Can you post a small code snippet of your `authExchange` config please? I'm confident I can debug this, but I'd need to see whether it's a mistake in the config first. 🤔 It's possible it's misidentifying a mutation response as an auth error and is sending a second one possibly.

**Reply by jbalke** [reply]
No problem and thanks for looking into it! 

Probably not relevant but my refresh token is stored in a HTTP-only cookie so refreshAccessToken() handles the POST request for a new access token. My access token is stored in memory. I've been trying to adhere to the following guidence: [https://hasura.io/blog/best-practices-of-using-jwt-with-graphql/](url).

    authExchange<{ token: string }>({
            addAuthToOperation: ({ authState, operation }) => {
              if (!authState || !authState.token) {
                return operation;
              }
              const fetchOptions =
                typeof operation.context.fetchOptions === 'function'
                  ? operation.context.fetchOptions()
                  : operation.context.fetchOptions || {};
              return {
                ...operation,
                context: {
                  ...operation.context,
                  fetchOptions: {
                    ...fetchOptions,
                    headers: {
                      ...fetchOptions.headers,
                      Authorization: `Bearer ${authState.token}`,
                    },
                  },
                },
              };
            },
            willAuthError: ({ authState }) => {
              if (!authState || isAccessTokenExpired()) return true;
              return false;
            },
            didAuthError: ({ error }) => {
              return error.graphQLErrors.some((e) => e.extensions?.code === 'UNAUTHENTICATED' );
            },
            getAuth: async ({ authState, mutate }) => {
              if (!authState) {
                let token = getAccessToken() || (await refreshAccessToken());
                if (token) {
                  return { token };
                }
                return null;
              }

              const newAccessToken = await refreshAccessToken();
              if (newAccessToken) {
                setAccessToken(newAccessToken);
                return {
                  token: newAccessToken,
                };
              }

              // otherwise, if refresh fails, log clear storage and log out
              clearAccessToken();
              await mutate('
                 mutate Logout {
                   logout
                 }
              ');
              Router.push('/login');

              return null;
            },
          })

**Reply by kitten** [reply]
is the argument-less `isAccessTokenExpired()`  call intentional?

**Reply by jbalke** [reply]
Yes, the access token is stored in a module-scoped variable, so enclosed by the utility functions (or was my intention 🤔)

**Reply by kitten** [reply]
Hm, I believe that would be alright, but you’ll have to be careful with race and sync conditions. Since the token is also stored in the `authState` your `isAccessTokenExoired` function will basically trigger `getAuth` to fetch a new one and sync it back. But again, that should be alright;

Does the problem disappear if `willAuthError` always returns `false` however?

**Reply by jbalke** [reply]
Ok, when `willAuthError` only returns `false` the issue doesn't occur.

So I reverted the change and amended `isAccessTokenExpired` to accept a token and not refer to the module-scoped variable, but the issue persists. This change is still in-place.

From what I've observed from a cold login, a scenario where my login mutation fires twice, this is the flow (a successful login will return an access token and set the refresh cookie):

    willAuthError -> true
    --authState is null

    getAuth - initialise
    --NO ACCESS TOKEN 
 
(not yet have http-only refresh token cookie)

    addAuthToOperation -> operation
    --no authState
    addAuthToOperation -> operation
    --no authState

  (a login mutation completes, access token set, refresh http-only cookie set)

    willAuthError -> true
    --authState is null

    getAuth - initialise
    --GOT ACCESS TOKEN

    addAuthToOperation -> operation with fetchOptions
    addAuthToOperation -> operation with fetchOptions

I'm still digging around but thought you might see something I've yet to wrap my head around...

**Reply by kitten** [reply]
Ok, I may have an idea of what's going on. I currently suspect that this call is duplicating the `operation` that has refreshed the authentication state: https://github.com/FormidableLabs/urql/blob/a770f18005c395245725e853d36246111591c37f/exchanges/auth/src/authExchange.ts#L155

**Comment by kitten** [comment, marked as answer]
I've just opened a PR to fix this issue and I believe I've tracked down what's causing this. https://github.com/FormidableLabs/urql/pull/1075

If you don't mind, you can test this out for me by installing it from Codesandbox CI ❤️ Basically just set the version in your `package.json` for `@urql/exchange-auth` to `"https://pkg.csb.dev/FormidableLabs/urql/commit/791f05aa/@urql/exchange-auth"`

**Reply by jbalke** [reply]
Looks like you got it, no more duplication. Nice one!

**Reply by kitten** [reply]
Awesome! This has also just been released :) https://github.com/FormidableLabs/urql/releases/tag/%40urql%2Fexchange-auth%400.1.1
